File: mvvm/JetLiYu/app/src/main/java/pri/weiqiang/jet/liyujapanese/data/database/VocabDatabase.java
package pri.weiqiang.jet.liyujapanese.data.database;

import android.content.Context;
import android.util.Log;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import pri.weiqiang.jet.liyujapanese.data.bean.Lesson;
import pri.weiqiang.jet.liyujapanese.data.bean.Word;
import pri.weiqiang.jet.liyujapanese.data.dao.LessonDao;
import pri.weiqiang.jet.liyujapanese.data.dao.WordDao;

@Database(entities = {Word.class,Lesson.class}, version = 1)
public abstract class VocabDatabase extends RoomDatabase {

    private static final String LOG_TAG = VocabDatabase.class.getSimpleName();
    private static final String DATABASE_NAME = "vocab";

    // For Singleton instantiation
    private static final Object LOCK = new Object();
    private static volatile VocabDatabase sInstance;

    public static VocabDatabase getInstance(Context context) {
        Log.d(LOG_TAG, "Getting the database");
        if (sInstance == null) {
            synchronized (LOCK) {
                sInstance = Room.databaseBuilder(context.getApplicationContext(),
                                VocabDatabase.class, VocabDatabase.DATABASE_NAME)
                        .createFromAsset("databases/vocab.db")
                        .build();
                Log.d(LOG_TAG, "Made new database");
            }
        }
        return sInstance;
    }

    public abstract WordDao wordDao();

    public abstract LessonDao lessonDao();
}

File: mvvm/KtJetLiYu/app/src/main/java/pri/weiqiang/jet/kt/liyujapanese/data/database/VocabDatabase.kt
package pri.weiqiang.jet.kt.liyujapanese.data.database

import android.content.Context
import android.util.Log
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import pri.weiqiang.jet.kt.liyujapanese.data.bean.Word
import pri.weiqiang.jet.kt.liyujapanese.data.dao.WordDao

@Database(entities = [Word::class], version = 1)
abstract class VocabDatabase : RoomDatabase() {
    abstract fun wordDao() : WordDao?

    companion object{
        private val DATABASE_NAME = "vocab"
        private val TAG = VocabDatabase::class.java.simpleName
        private var sInstance:VocabDatabase? = null
        fun getInstance(context:Context):VocabDatabase?{
//            return sInstance ?: synchronized(this){
//                sInstance = Room.
//                    databaseBuilder(context.applicationContext,VocabDatabase::class.java, DATABASE_NAME)
//                    .createFromAsset("databases/vocab.db")
//                    .build()
//                return sInstance
//            }
            if (sInstance == null) {
                synchronized(this) {
                    sInstance = Room.databaseBuilder(
                        context.applicationContext,
                        VocabDatabase::class.java, DATABASE_NAME
                    )
                        .createFromAsset("databases/vocab.db")
                        .build()
                    Log.d(TAG, "Made new database")
                }
            }
            return sInstance
        }
    }
}